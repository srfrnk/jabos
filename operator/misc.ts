import { ApiObjectProps, Chart } from 'cdk8s';
import { GitRepositoryProps } from './imports/jabos.io';
import { Container, KubeJob, KubeJobProps, KubeRoleBinding, Quantity } from './imports/k8s';
import { SyncRequest } from './metaControllerHooks';

export function k8sName(prefix: string, commit: string): string {
  return `${prefix.substring(0, 52)}-${commit}`.substring(0, 62);
}

export interface GitRepositoryPropsEx extends GitRepositoryProps {
  readonly status: {
    readonly latestCommit: string,
  }
}

export function getRepo(request: SyncRequest): GitRepositoryPropsEx {
  const namespace = request.object.metadata.namespace;
  const reposMap = (request.related || {})['GitRepository.jabos.io/v1'] /* as { [key: string]: GitRepositoryProps } */;
  var repo: GitRepositoryProps;

  // Have to filter, due to the following bug: https://github.com/metacontroller/metacontroller/issues/414
  const repos = Object.values(reposMap || {}).filter((r) => r?.metadata?.namespace === namespace);
  if (repos.length > 0) {
    repo = repos[0] as unknown as GitRepositoryProps;
  }
  else {
    // Have to reject the sync to maintain idempotent response - due to: https://github.com/metacontroller/metacontroller/issues/414
    throw new Error(`No GitRepository from same namespace.`);
  }

  return {
    status: { latestCommit: '' },
    spec: { url: '', branch: '', ssh: null },
    ...repo,
  };
}

function convertQuantities(resources: { [key: string]: (Quantity | string | number) }): { [key: string]: Quantity } {
  return Object.entries(resources).reduce((p, c) => {
    if (typeof c[1] === 'string') {
      p[c[0]] = Quantity.fromString(c[1]);
    }
    else if (typeof c[1] == 'number') {
      p[c[0]] = Quantity.fromNumber(c[1]);
    }
    else {
      p[c[0]] = c[1];
    }
    return p;
  }, {});
}

function convertContainers(containers: Container[]) {
  return containers.map(container => {
    return {
      ...container,
      ...(container.resources && {
        resources: {
          ...(container.resources.limits && { limits: convertQuantities(container.resources.limits) }),
          ...(container.resources.requests && { requests: convertQuantities(container.resources.requests) })
        }
      })
    };
  });
}

export function useExistingJob(chart: Chart, request: SyncRequest) {
  if (Object.values(request.attachments['Job.batch/v1']).length > 0) {
    var jobProps: KubeJobProps = Object.values(request.attachments['Job.batch/v1'])[0];
    jobProps = {
      ...jobProps,
      ...(jobProps.metadata && {
        metadata: {
          ...jobProps.metadata,
          ...(jobProps.metadata.annotations && {
            annotations: {
              ...jobProps.metadata.annotations,
              'metacontroller.k8s.io/last-applied-configuration': null
            },
          }),
        }
      }),
      ...(jobProps.spec && {
        spec: {
          ...jobProps.spec,
          ...(jobProps.spec.template && {
            template: {
              ...jobProps.spec.template,
              ...(jobProps.spec.template.spec && {
                spec: {
                  ...jobProps.spec.template.spec,
                  ...(jobProps.spec.template.spec.initContainers && { initContainers: convertContainers(jobProps.spec.template.spec.initContainers) }),
                  ...(jobProps.spec.template.spec.containers && { containers: convertContainers(jobProps.spec.template.spec.containers) }),
                }
              })
            }
          })
        }
      })
    };
    return new KubeJob(chart, jobProps.metadata.name, jobProps);
  }
}

export function debugId(object: ApiObjectProps) {
  return `${object?.metadata?.namespace}:${object?.metadata?.name}`;
}

export function needNewBuild(request: SyncRequest): boolean {
  const object = request.object;
  const repo = getRepo(request);
  const existingJob = Object.values(request.attachments['Job.batch/v1']).length > 0;
  const builtCommit: string = (object.status || {}).builtCommit || '';
  const latestCommit: string = repo.status.latestCommit;
  return !existingJob && !!latestCommit && latestCommit !== builtCommit;
}

export const _private = { convertQuantities: convertQuantities };
