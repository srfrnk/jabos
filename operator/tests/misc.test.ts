import { ApiObjectProps } from 'cdk8s';
import { Quantity } from '../imports/k8s';
import { debugId, getRepo, _private } from '../misc';

test('misc getRepo no related', () => {
  expect(() => {
    getRepo({
      object: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: undefined,
      finalizing: false,
      controller: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
      }
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo missing repo', () => {
  expect(() => {
    getRepo({
      object: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          undefined as ApiObjectProps
        ]
      },
      finalizing: false,
      controller: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
      }
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo missing metadata', () => {
  expect(() => {
    getRepo({
      object: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          {
            apiVersion: 'apiVersion_value',
            kind: 'kind_version',
          }
        ]
      },
      finalizing: false,
      controller: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
      }
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo missing namespace', () => {
  expect(() => {
    getRepo({
      object: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          {
            apiVersion: 'apiVersion_value',
            kind: 'kind_version',
            metadata: {
            }
          }
        ]
      },
      finalizing: false,
      controller: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
      }
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo bad namespace', () => {
  expect(() => {
    getRepo({
      object: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          {
            apiVersion: 'apiVersion_value',
            kind: 'kind_version',
            metadata: {
              namespace: 'bad_namespace_value'
            }
          }
        ]
      },
      finalizing: false,
      controller: {
        apiVersion: 'apiVersion_value',
        kind: 'kind_version',
      }
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo no spec and status', () => {
  expect(getRepo({
    object: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
      metadata: {
        namespace: 'namespace_value'
      }
    },
    attachments: [],
    related: {
      'GitRepository.jabos.io/v1': [
        {
          apiVersion: 'apiVersion_value',
          kind: 'kind_version',
          metadata: {
            namespace: 'namespace_value'
          },
        }
      ]
    },
    finalizing: false,
    controller: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
    }
  })).toMatchSnapshot();
});

test('debugId no value', () => {
  expect(debugId({} as ApiObjectProps)).toMatchSnapshot();
});

test('debugId undefined', () => {
  expect(debugId(undefined as ApiObjectProps)).toMatchSnapshot();
});

test('convertQuantities with number', () => {
  expect(_private.convertQuantities({ num: 10 })).toMatchSnapshot();
});

test('convertQuantities with Quantity', () => {
  expect(_private.convertQuantities({ num: Quantity.fromNumber(10) })).toMatchSnapshot();
});
