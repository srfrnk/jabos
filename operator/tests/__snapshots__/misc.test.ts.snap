// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertQuantities with Quantity 1`] = `
Object {
  "num": Quantity {
    "value": 10,
  },
}
`;

exports[`convertQuantities with number 1`] = `
Object {
  "num": Quantity {
    "value": 10,
  },
}
`;

exports[`debugId no value 1`] = `"undefined:undefined"`;

exports[`debugId undefined 1`] = `"undefined:undefined"`;

exports[`misc getRepo no spec and status 1`] = `
Object {
  "apiVersion": "apiVersion_value",
  "kind": "kind_version",
  "metadata": Object {
    "namespace": "namespace_value",
  },
  "spec": Object {
    "branch": "",
    "ssh": null,
    "url": "",
  },
  "status": Object {
    "latestCommit": "",
  },
}
`;
